File: server/src/main/resources/static/Socket.js
import {args} from "./utils.js";

let globalSocket = null;
let globalCallbacks = [];

export class Socket {
    eventListeners = new Map();
    connectListeners = new Set();
    disconnectListeners = new Set();
    errorListeners = new Set();
    socket;

    constructor(url, params = {}) {
        this.socket = this._connect(url, params);
    }

    _connect(url, params = {}) {
        const socket = new WebSocket(`${location.protocol.replace("http", "ws")}//${location.host}${url}${args(params)}`);
        socket['onopen'] = (e) => {
            this.connectListeners.forEach(it => it(e));
        };
        socket['onclose'] = (e) => {
            this.disconnectListeners.forEach(it => it(e));
            setTimeout(() => this._connect(url, params), 5000);
        };
        socket['onerror'] = (err) => {
            this.errorListeners.forEach(it => it(err));
        };
        socket['onmessage'] = (e) => {
            console.log("in", e);
            let json = JSON.parse(e.data);
            this.eventListener(json.type).forEach(it => it(json));
        };
    }

    send(json) {
        this.socket.send(JSON.stringify(json));
    }

    eventListener(key) {
        if (!this.eventListeners.has(key)) {
            this.eventListeners.set(key, new Set());
        }
        return this.eventListeners.get(key);
    }

    discover(x, y) {
        this.send({type: 'discover', x, y});
    }

    static start(url, params) {
        globalSocket = new Socket(url, params);
        globalCallbacks.forEach(it => it(globalSocket));
    }

    static socket(callback) {
        if (globalSocket != null) {
            callback(globalSocket);
        } else {
            globalCallbacks.push(callback);
        }
    }
}

File: js/src/main/kotlin/Socket.kt
import kotlinx.serialization.json.Json
import org.w3c.dom.WebSocket
import org.w3c.dom.events.Event
import ru.o4fun.models.Incoming
import kotlin.browser.window
import kotlin.reflect.KClass

class Socket(base: String, params: Map<String, Any> = mapOf()) {
    private var socket : WebSocket? = null
    private val url = "${window.location.protocol.replace("http", "ws")}//${window.location.host}${base}${args(params)}"
    private val eventListeners = mutableMapOf<KClass<out Incoming>, MutableSet<(Incoming) -> Unit>>()
    val connectListeners = mutableSetOf<(Event) -> Unit>()
    val disconnectListeners = mutableSetOf<(Event) -> Unit>()
    val errorListeners = mutableSetOf<(Event) -> Unit>()

    init {
        connect()
    }

    private fun connect() {
        socket = WebSocket(url).apply {
            onopen = { e ->
                connectListeners.forEach { it(e) }
            }
            onmessage = { e ->
                console.log("in", e)
                val json = Json.parse(
                    Incoming.serializer(),
                    e.data.toString()
                )
                eventListener(json::class).forEach { it(json) }
            }
            onclose = { e ->
                disconnectListeners.forEach { it(e) }
                window.setTimeout({
                    connect()
                }, 5000)
            }
            onerror = { err ->
                errorListeners.forEach { it(err) }
            }
        }
    }

    fun send(json: Incoming) = socket?.send(
        Json.stringify(
            Incoming.serializer(),
            json
        )
    )

    fun eventListener(key: KClass<out Incoming>) = eventListeners.getOrPut(key) { mutableSetOf() }

    fun discover(x: Int, y: Int) = send(Incoming.Discover(x, y))

    companion object {
        var socket : Socket? = null
        val callbacks = mutableSetOf<(Socket) -> Unit>()

        fun start(url : String, params: Map<String, Any>) {
            socket = Socket(url, params)
            callbacks.forEach { it(socket!!) }
        }

        fun socket(callback: (Socket) -> Unit) {
            socket?.let {callback(it)} ?: callbacks.add(callback)
        }
    }
}